[Quantizer][Add] Fix add annotation with constant

mcr229 opened the pull request:
Summary:
Occaisonally we run into a partition that looks like this for Add:

```
SourcePartition(nodes=[_constant2, add_2], source=<built-in function add>, input_nodes=[x], output_nodes=[_constant2, add_2], params=[_constant2])
```

In this case we are adding a constant to an input, and reusing the constant later down the line. This causes our constant to be an output in our SourcePartition. The assumption then that:

```
        add_node = add_partition.output_nodes[0]
```
Will not necessarily hold. As a result we must check that the output node is indeed a call function and not a constant.

Test Plan: buck test mode/dev-nosan //executorch/backends/xnnpack/test:test_xnnpack_ops -- test_qs8_add_constant

Differential Revision: D60413221

jerryzh168 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -825,6 +825,12 @@ def _annotate_add(
+        for output_node in add_partition.output_nodes:
+            # constants may be propagated through partition as an
+            # output, so we need to check that output node is a call_function
+            if output_node.op == "call_function":
```
I see, this could be true for mul patterns as well. thoughts on refactoring the matching code to use `aten.add`/`aten.add_` ops? cc @kimishpatel

jerryzh168 reviewed (approved):
stamping to unblock first

mcr229 committed:
[Quantizer][Add] Fix add annotation with constant (#132092)

Summary:
Occaisonally we run into a partition that looks like this for Add:

```
SourcePartition(nodes=[_constant2, add_2], source=<built-in function add>, input_nodes=[x], output_nodes=[_constant2, add_2], params=[_constant2])
```

In this case we are adding a constant to an input, and reusing the constant later down the line. This causes our constant to be an output in our SourcePartition. The assumption then that:

```
        add_node = add_partition.output_nodes[0]
```
Will not necessarily hold. As a result we must check that the output node is indeed a call function and not a constant.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/132092

Test Plan: buck test mode/dev-nosan //executorch/backends/xnnpack/test:test_xnnpack_ops -- test_qs8_add_constant

Reviewed By: jerryzh168

Differential Revision: D60413221

Pulled By: mcr229
